autoprecompiles/src/optimizer.rs

```diff
@@ -0,0 +1,182 @@
+            .map(|constraint| algebraic_to_quadratic_symbolic_expression(&constraint.expr))
+            .collect(),
+        bus_interactions: symbolic_machine
+            .bus_interactions
+            .iter()
+            .map(symbolic_bus_interaction_to_bus_interaction)
+            .collect(),
+    }
+}
+
+fn constraint_system_to_symbolic_machine<P: FieldElement>(
+    constraint_system: ConstraintSystem<P, Variable>,
+) -> SymbolicMachine<P> {
+    SymbolicMachine {
+        constraints: constraint_system
+            .algebraic_constraints
+            .iter()
+            .map(|constraint| SymbolicConstraint {
+                expr: simplify_expression(quadratic_symbolic_expression_to_algebraic(constraint)),
+            })
+            .collect(),
+        bus_interactions: constraint_system
+            .bus_interactions
+            .into_iter()
+            .map(bus_interaction_to_symbolic_bus_interaction)
+            .collect(),
+    }
+}
+
+fn solver_based_optimization<T: FieldElement>(
+    constraint_system: ConstraintSystem<T, Variable>,
+) -> ConstraintSystem<T, Variable> {
+    let result = Solver::new(constraint_system)
+        .solve()
+        .map_err(|e| {
+            panic!("Solver failed: {e:?}");
+        })
+        .unwrap();
+    log::trace!("Solver figured out the following assignments:");
+    for (var, value) in result.assignments.iter() {
+        log::trace!("  {var} = {value}");
+    }
+    result.simplified_constraint_system
+}
+
+fn remove_trivial_constraints<P: FieldElement>(
+    mut symbolic_machine: ConstraintSystem<P, Variable>,
+) -> ConstraintSystem<P, Variable> {
+    let zero = QuadraticSymbolicExpression::from(P::zero());
+    symbolic_machine
+        .algebraic_constraints
+        .retain(|constraint| constraint != &zero);
+    symbolic_machine
+        .bus_interactions
+        .retain(|bus_interaction| bus_interaction.multiplicity != zero);
+    symbolic_machine
+}
+
+fn symbolic_bus_interaction_to_bus_interaction<P: FieldElement>(
+    bus_interaction: &SymbolicBusInteraction<P>,
+) -> BusInteraction<QuadraticSymbolicExpression<P, Variable>> {
+    BusInteraction {
+        bus_id: SymbolicExpression::Concrete(P::from(bus_interaction.id)).into(),
+        payload: bus_interaction
+            .args
+            .iter()
+            .map(|arg| algebraic_to_quadratic_symbolic_expression(arg))
+            .collect(),
+        multiplicity: algebraic_to_quadratic_symbolic_expression(&bus_interaction.mult),
+    }
+}
+
+fn bus_interaction_to_symbolic_bus_interaction<P: FieldElement>(
+    bus_interaction: BusInteraction<QuadraticSymbolicExpression<P, Variable>>,
+) -> SymbolicBusInteraction<P> {
+    // We set the bus_id to a constant in `bus_interaction_to_symbolic_bus_interaction`,
+    // so this should always succeed.
+    let id = bus_interaction
+        .bus_id
+        .try_to_number()
+        .unwrap()
+        .to_arbitrary_integer()
+        .try_into()
+        .unwrap();
```

Comment:
Rust + our APIs lol

Reply:
There us `to_degree()`, but Chris told me not to use it (which makes sense, "degree" is really semantically wrong here). There might be a more ergonomic way, but then it's not obvious.